### PR TITLE
Enable parallel OpenAI calls

### LINKED ISSUE #493: openai 1.x generator doesn't work nice with `parallel_attempts`
see
```
(garak) 10:48:53 x1:~/dev_ext/garak [feature/oai_model_update] $ python3 -m garak -p dan.DanInTheWild -m openai -n gpt-4-turbo-preview -g 1 --parallel_attempts 20
garak LLM security probe v0.9.0.11.post2 ( https://github.com/leondz/garak ) at 2024-02-20T10:49:00.692318
📜 reporting to garak_runs/garak.cb7f7e26-3bc0-45b3-a95c-4615a6685ce2.report.jsonl
🦜 loading generator: OpenAI: gpt-4-turbo-preview
🕵️  queue of probes: dan.DanInTheWild
probes.dan.DanInTheWild:   0%|                                                                                                                                                            | 0/666 [00:00<?, ?it/s]Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/lderczynski/dev_ext/garak/garak/__main__.py", line 13, in <module>
    main()
  File "/home/lderczynski/dev_ext/garak/garak/__main__.py", line 9, in main
    cli.main(sys.argv[1:])
  File "/home/lderczynski/dev_ext/garak/garak/cli.py", line 477, in main
    command.probewise_run(generator, probe_names, evaluator, buffs)
  File "/home/lderczynski/dev_ext/garak/garak/command.py", line 216, in probewise_run
    probewise_h.run(generator, probe_names, evaluator, buffs)
  File "/home/lderczynski/dev_ext/garak/garak/harnesses/probewise.py", line 108, in run
    h.run(model, [probe], detectors, evaluator, announce_probe=False)
  File "/home/lderczynski/dev_ext/garak/garak/harnesses/base.py", line 92, in run
    attempt_results = probe.probe(model)
                      ^^^^^^^^^^^^^^^^^^
  File "/home/lderczynski/dev_ext/garak/garak/probes/base.py", line 131, in probe
    for result in attempt_pool.imap_unordered(
  File "/home/lderczynski/anaconda3/envs/garak/lib/python3.12/multiprocessing/pool.py", line 873, in next
    raise value
  File "/home/lderczynski/anaconda3/envs/garak/lib/python3.12/multiprocessing/pool.py", line 540, in _handle_tasks
    put(task)
  File "/home/lderczynski/anaconda3/envs/garak/lib/python3.12/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lderczynski/anaconda3/envs/garak/lib/python3.12/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle '_thread.RLock' object
```

contributing factors:
* `openai` module 1.x contains `_thread.RLock`
* `multiprocessing` uses `pickle`
* `pickle` can't handle `_thread.RLock`
* `base.probe()` links the generator in `self.generator`

possible mitigations:
* get the openai generator to work out if its being pickled, and avoid that
* use something other than `pickle`
* don't try to pickle the generator
* don't try to pickle this generator - maybe have a flag for generators that support parallelism (and hope these don't go stale? maybe a test that looks for exceptions in generators that are tagged non-parallelisable?)

possible upgrades:
* detect parallelisation failures and fall back to no parallelism, instead of exploding

### PR DESCRIPTION
**Pull request by jmartin-tech**
Fix #493 

Conceptually since rest generator client objects should not retains state, this provides a pattern where `rest` generator client objects can be instantiated only when required and avoiding the need to `pickle` the client object when passing a generator to another process.

* refactor for just in time rest client for `OpenAI`
* ensure read and write of `_config.transient.reportfile` are in parent process only
* consolidate multiprocessing execution in `probes/base.py`

Verification testing preformed with:
```
python -m garak p dan.DanInTheWild -m openai -n gpt-4-turbo-preview -g 1 --parallel_attempts 20
```

Additional testing may be appropriate to validate changes to `base` and `gcg` probe flows are not impacted.

### CODE REVIEW

**Review by erickgalinkin — APPROVED** [review]
Looks great!

**Comment by erickgalinkin** [comment]
I did not test, only performed code review. But nothing here seems to remove or alter functionality in a breaking way.

**Comment by jmartin-tech** [comment]
Extended refactor to provide example of using a base class that can share `OpenAI` compatible generation.

File: garak/generators/openai.py [hunk 1]
```diff
@@ -82,58 +82,62 @@
...
 
-        api_key = os.getenv("OPENAI_API_KEY", default=None)
-        if api_key is None:
+        self.api_key = os.getenv(self.ENV_VAR, default=None)
+        if self.api_key is None:
             raise ValueError(
-                'Put the OpenAI API key in the OPENAI_API_KEY environment variable (this was empty)\n \
-                e.g.: export OPENAI_API_KEY="sk-123Xgenerators/openai.pyXXXXXXXXXXX"'
+                f'Put the {self.generator_family_name} API key in the {self.ENV_VAR} environment variable (this was empty)\n \
+                e.g.: export {self.ENV_VAR}="sk-123Xgenerators/openai.pyXXXXXXXXXXX"'
```
Comment (leondz) [inline garak/generators/openai.py:118]: weird old paste typo
```suggestion
                e.g.: export {self.ENV_VAR}="sk-123XXXXXXXXXXXX"'
```

File: garak/probes/base.py [hunk 2]
```diff
@@ -139,54 +139,62 @@ def _execute_attempt(self, this_attempt):
...
-        prompts = list(self.prompts)
-        for seq, prompt in enumerate(prompts):
-            attempts_todo.append(self._mint_attempt(prompt, seq))
-
-        # buff hook
-        if len(_config.buffmanager.buffs) > 0:
-            attempts_todo = self._buff_hook(attempts_todo)
-
-        # iterate through attempts
+    def _execute_all(self, attempts) -> Iterable[garak.attempt.Attempt]:
```
Comment (leondz) [inline garak/probes/base.py:145]: can `docs/source/garak.probes.base.rst` get a small update to document this change?

File: garak/probes/dan.py [hunk 3]
```diff
@@ -607,33 +606,7 @@ def probe(self, generator) -> Iterable[garak.attempt.Attempt]:
...
-                            result
-                        )  # these will be out of original order
-                        attempt_bar.update(1)
-
-            else:
-                attempt_iterator = tqdm(attempts_todo, leave=False)
-                attempt_iterator.set_description(self.probename.replace("garak.", ""))
-                for this_attempt in attempt_iterator:
-                    attempts_completed.append(self._execute_attempt(this_attempt))
+            attempts_completed = self._execute_all(attempts_todo)
```
Comment (leondz) [inline garak/probes/dan.py:609]: :)

File: garak/probes/gcg.py [hunk 4]
```diff
@@ -98,33 +96,7 @@ def probe(self, generator) -> List[garak.attempt.Attempt]:
...
-                            result
-                        )  # these will be out of original order
-                        attempt_bar.update(1)
-
-            else:
-                attempt_iterator = tqdm.tqdm(attempts_todo, leave=False)
-                attempt_iterator.set_description(self.probename.replace("garak.", ""))
-                for this_attempt in attempt_iterator:
-                    attempts_completed.append(self._execute_attempt(this_attempt))
+            attempts_completed = self._execute_all(attempts_todo)
```
Comment (leondz) [inline garak/probes/gcg.py:100]: :) :)

**Review by leondz — APPROVED** [review]
lgtm - would just appreciate one doc update

File: garak/probes/base.py (hunk 2 above)
Comment (jmartin-tech) [reply garak/probes/base.py:145]: New commit c5eaf39 adds doc changes, I am not 100% sure on formatting or content so please offer any suggestions you believe are needed still.

Comment (leondz) [reply garak/probes/base.py:145]: I think this is fine!

**Pull request merged**